[ Feature ] Header 제작

Jaeppetto opened the pull request:
## 📝 설명

Header 컴포넌트를 작업했습니다.

## ✅ PR 유형

<!--
	팀원이 어떤 작업을 하였는지 쉽게 알아볼 수 있게 도와주는 부분입니다.
-->

- [x] 새로운 기능 추가
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드 리팩토링
- [ ] 파일 혹은 폴더명 수정 / 삭제
- [ ] 기타

## 💻 작업 내용

- 현재 Segment(URL)을 기준으로 알맞은 State를 반환하는 `useSegmentConvert` 훅을 구현하였습니다.
- 해당 훅을 통해 반환받은 State (`visible`, `headerLabel`, `leftButtonIcon`, `leftButtonFn,` `rightButtonIcon`, `rightButtonFn`)를 이용해 화면에 렌더링합니다.
- Tailwind Class Sorter 디펜던시를 설치하였습니다. Dev 브랜치 반영 후 `yarn install` 한 번씩 부탁드립니다!
- Segment를 반환하는 `useSelectedLayoutSegments` 훅이 현재 URL을 기준으로 작동하는 관계로, 별도의 스토리북 컴포넌트는 생성하지 않았습니다.

<!-- PR 본문을 입력하세요. -->

## 💬 PR 포인트

<!-- PR 리뷰 시 공유 사항 또는 유심히 보면 좋을 부분을 설명합니다. -->

### 컴포넌트 구조와 사용방법

1. Header 컴포넌트내에서 `useSelectedLayoutSegments` 훅을 이용해 현재 Segments를 불러옵니다. Root Layout 기준이며, 만약 현재 /home/user/id의 경로에 위치할 경우 ['home', 'user', 'id']의 배열을 `segments` 변수에 저장하게 됩니다.
2. 이 segments를 `useSegmentConvert` 커스텀훅에 전달합니다. 이를 전달받아 해당 segments에 맞는 state를 반환하게 됩니다. 따라서 사용하실 때, **`useSegmentConvert` 훅 내부에서 switch 문을 이용해 분기할 경로를 추가하시고, state를 변환해주시면 됩니다!**
3. 설정해주신 값에 맞게, 해당 URL에서 헤더가 출력됩니다.
4. **주의하실 점**은 `leftButtonIcon`, `rightButtonIcon`을 설정하실 때, 기존 아이콘을 state로 넘겨주면 안된다는 점입니다! 기존 아이콘은 Prop을 가지고 있는 관계로 state 값에 담겨지지가 않는데요, 따라서 Props를 가지고 있지 않은 아이콘의 크기를 직접 설정하신 후 import하여 적용해주서야 합니다! (/Header/staticIcons/ArrowLeftIcon.tsx 를 참고하시면 될 것 같습니다.)


아래는 헤더 컴포넌트 이용 간 조작하실 `useSegmentConvert` 훅에 대한 구조적인 설명입니다. 추가적으로 궁금하신 점 있으시면 언제든 여쭤보셔도 됩니다!

<img width="849" alt="스크린샷 2024-05-03 오후 5 51 55" src="https://github.com/DDu-Du-DDu-Du/DDuDu_Front/assets/95916813/6a93c2ee-bd3c-4245-8f88-8f0b367c4b76">
<img width="484" alt="스크린샷 2024-05-03 오후 4 45 34" src="https://github.com/DDu-Du-DDu-Du/DDuDu_Front/assets/95916813/aef744b8-9017-49a2-945a-fbfa2a35bb82">

## 🚀 피드백 반영사항
1. 헤더가 사용되는 페이지들을 그룹 라우트로 분리하였습니다. `(header)` 내부에 라우트 폴더를 생성하시면 됩니다! (9d2776df5fcf05960f5cdda905687d1942ce4a85)
2. useEffect, useState로 상태를 관리하던 기존의 useSegmentConvert 훅의 로직을 useMemo로 대체하였습니다. 그러나 복잡하지 않은 로직을 메모이제이션하는 것은 추가적인 오버헤드를 발생시킬 수도 있기 때문에, 추후 수치적인 검증이 필요할 것으로 생각됩니다.(342dc625c77e16218a3a70c34c05bf9f3d7d30ed)
3. useSegmentConvert 훅을 useHeaderLabel, useHeaderRightButton 훅으로 분리하였습니다. (a18c1b5b8264ac05e04dbc1a9131a210c66faf01)

Jaeppetto committed:
💎feat: 헤더 기능 구현

- 현재 Segment(URL)을 기준으로 알맞은 State를 반환하는 useSegmentConvert 훅
- 반환받은 State를 화면에 출력하는 Header 컴포넌트
- add tailwindCSS Sort Dependency

Jaeppetto committed:
💎feat: leftButtonIcon 기본 아이콘 추가

Jaeppetto committed:
💎feat: 헤더 레이블 클릭 시 스크롤 최상단 이동

Jae0o commented on src/app/_components/client/Header/hooks/useSegmentConvert/useSegmentConvert.ts:
```diff
@@ -0,0 +1,46 @@
+
+import { ReactNode, useEffect, useState } from "react";
+
+import ArrowLeftIcon from "../../staticIcons/ArrowLeftIcon/ArrowLeftIcon";
```
[ P2 ] 
Icon은 _components/sever/Icons에서 관리하기로 하기로 했던걸로 기억하는데 맞다면 거기서 다같이 관리하는게 좋을것같습니다!

Jaeppetto replied:
Props가 없는 아이콘들이라 헤더에서만 쓰일 것 같아 따로 빼놓았었는데, 추후 다른 곳에서 쓰일 수도 있으니 /Icons에서 관리하는게 좋을 것 같네요! 반영하겠습니다 👍🏻

Jae0o commented on src/app/_components/client/Header/hooks/useSegmentConvert/useSegmentConvert.ts:
[ P5 ]
개인적인 제안을 몇가지 드려보고싶은데요!
그전에 일단 개인적으로 
1. Header layout이나 특정 부분을 RSC로 만들 방법은 없을까 라는 생각이 살짝 드네요!
  그럼 좋겠다라고 생각한 부분은 저번에 말씀드린 부분처럼 Header는 SSG를 사용가능한 정적인 페이지가 될 수 있을것 같다는 생각이 들어서 방법이 없을까 라는 생각을 하게되었습니다.

2. 아래 제작해주신 로직에 문제가 있는건 아니지만 현재 각각의 useState들이 존재하고 url에 따라 useEffect내부에서 switch문을 통해 제어가 되어지고 있는것같은데 해당 부분을 useMemo를 통해 제어하는 방법은 어떨까 라는 생각을 좀 하게되었습니다.
```
const useHeaderLeftButton = (segment) => {
  return useMemo(() => {
    switch(segment){
      case "main" :
        return 메인에 사용될 아이콘

      case "user" :
        return 메인에 사용될 아이콘

      default :
        return 메인에 사용될 아이콘
    }
  },[segment])
}

const useHeaderRightButton = (segment) => {
  const rightButton useMemo(() => {
    switch(segment){
      case "main" :
        return 메인에 사용될 아이콘

      case "user" :
        return 메인에 사용될 아이콘

      default :
        return 메인에 사용될 아이콘
    }
  },[segment])
  
  return rightButton
}
```
이런식으로 관리한다면 확실히 구역?  혹은 별도의 의도별로 각각의 hook을 제작할 수 있어 저희의 custom hook방식에 더 맞지 않을까 생각해봤어요.
일단 useEffect의 실행 타이밍이 보장되지 않는 순간이 있다는 위험이 줄어들것 같기도 하네요!

SeungHyune replied:
[ ASK ]
지금 코드처럼 `useEffect`와 `useState`를 사용하는 것과 `useMemo`를 사용하는 것에 대한 **각각의 장단점이 있는 것 같은데** memo를 사용했을 때의 장점은 

1. 실행 시점을 보장할 수 있다.
2. state를 사용하지 않고 상태를 관리한다.

두 가지가 맞을까요 ??!

성능적인 측면해서도 이점이 있는건지 궁금하여 질문 드립니다 ㅎㅎ 😊

Jaeppetto replied:
@Jae0o
1. 재영님 말씀처럼 Header는 현재 `useSelectedLayoutSegments` 훅 때문에 클라이언트 훅으로 유지되고 있는데요, 좀 더 Atomic하게 요소들을 분리하면 충분히 서버 컴포넌트로도 적용이 가능하다고 생각됩니다. 각각의 하위 컴포넌트를 클라이언트 컴포넌트로 관리하면 되니까요!
2. (1)번의 답변처럼 컴포넌트들을 Atomic하게 분리한다면 훅 또한 기능별로 나누어 관리할 수 있을 것 같은데요, 우려되는 부분이 한 가지 있다면 새로운 라우트에 대한 **헤더를 설정할 때마다 다섯 개의 훅 파일을 돌아다니며 상태 갱신 함수를 작성**해주어야 한다는 점입니다 🥲

`커스텀훅`은 하나의 파일에서 관리하되, 좀 더 가독성이나 메모이제이션 부분을 고려하고 + `JSX컴포넌트`는 Atomic하게 분리해 원하는 상태값만 가져와 그려주는 방향으로 가는 건 어떨까요?

Jae0o replied:
훅 파일이 돌아다니는건 JS를 파싱하는 단계에서 차이가 존재하지 않기때문에 (정확히는 존재하더라도 0.0 몇초 차) 라고 알고 있어서 저는 개인적으로는 컨벤션과 일전에 저희끼리 말한 패턴에 맞게 좀 더 컴포넌트 단위도 잘게 나누고 훅도 의미대로 나누는쪽이 좀더 좋지 않을까 라는 생각이 들긴하네요

Jaeppetto replied:
아 제가 말씀드린 의도는 단순히 **너무 귀찮지 않을까요?!** 였습니다..ㅋㅋ 🤭 월요일 스크럼 때 다같이 얘기해봐도 좋을 것 같아요!

Jae0o commented on src/app/_components/client/Header/hooks/useSegmentConvert/useSegmentConvert.ts:
```diff
@@ -0,0 +1,46 @@
+    }
+  }, [segments]);
+
+  return [visible, headerLabel, leftButtonIcon, leftButtonFn, rightButtonIcon, rightButtonFn];
```
[ ASK ] 
오류 발생안했나요? 보통 배열의 경우 순서를 보장할 수 없어서 eslint에서 Error를 발생시키더라구요 드래서 객체로 내보내고있는데

Jaeppetto replied:
네 따로 린팅이나 Husky에서 에러는 발생하지 않았습니다만, 배열로 반환하면 그런 문제점이 있었군요! 반영하겠습니다 ㅎ_ㅎ

Jae0o replied:
lint가 문제를 일으키지는 않더라도 잘못된 사용을 방지하기 위해서라도 객체가 좋을것 같네요

Jae0o reviewed (approved):
첫 PR 수고하셨습니다~

SeungHyune commented on src/app/_components/client/Header/Header.tsx:
```diff
@@ -0,0 +1,48 @@
+
+const Header = () => {
+  const segments = useSelectedLayoutSegments();
+  const [visible, headerLabel, leftButtonIcon, leftButtonFn, rightButtonIcon, rightButtonFn] =
```
[ P5 ]
`useSegmentConvert hook`의 반환값을 {}로 변경하면 해당 반환 값들에 대한 타입도 추론이 될 것 같습니다 !!
아래 as 문은 모두 제거할 수 있겠네요 ㅎㅎ

Jaeppetto replied:
그래서 타입 에러가 발생했던거군요! 감사합니다 ☺️ 반영하겠습니다!

SeungHyune reviewed (approved):
재웅님 바쁘신 와중에 고생 많으셨습니다 !! 😊

Jaeppetto committed:
🪄change: StaticIcons 이동

Jaeppetto committed:
🛠️refactor: 컴포넌트 분리

- HeaderButton
- HeaderLabel

Songhee99 reviewed (approved):
재웅님 고생 많으셨습니다~~ 첫 pr부터 아주 야무집니다 👍🏻👍🏻
다른 분들이 질문을 잘 해주셔서 저도 생각해보게 되네요 ㅎㅎㅎㅎ

Jaeppetto committed:
🛠️refactor: Header를 사용하는 페이지 그룹 라우트로 분리

Jaeppetto committed:
🛠️refactor: useSegmentConvert 훅 로직 개선

- useEffect, useState로 구성된 기존 로직을 useMemo로 통합
- LeftIcon, LeftFunction 정적인 값 부여 (뒤로가기 버튼으로 고정)

Jaeppetto committed:
🪄change: useMemo 반환 형태 변경

Jaeppetto committed:
🪄change: Header 컴포넌트 index.ts 추가 및 아이콘 경로 수정

Jaeppetto committed:
🛠️refactor: useSegmentConvert 훅 분리, index.ts 수정
